### PR TITLE
SAMZA-2381: Refactor JobRunner#run and split into multiple util methods.

### PR DESCRIPTION
**Pull request by kw2542**
Issues:
  JobRunner#run is composed of three parts, persisting config in coordinator stream, creation of diagnostics stream when applicable and submit the job. It is not easy to read when all of them are combined in a single method and it cannot be reused when a subset of the functionality is needed elsewhere. For example, when we move planning to AM, we will only need to submit job on runner but the rest on AM.

Changes:
  1. Move coordinator stream related logic to CoordinatorStreamUtil#writeConfigToCoordinatorStream
  2. Move diagnostics stream related logic to DiagnosticsUtil#createDiagnosticsStream
  3. Move job submission related logic to JobRunner#submit()

Tests:
  1. Add unit tests to test CoordinatorStreamUtil#writeConfigToCoordinatorStream

### CODE REVIEW

File: samza-core/src/main/scala/org/apache/samza/job/JobRunner.scala [hunk 1]
```diff
@@ -66,17 +66,33 @@ object JobRunner extends Logging {
...
    * @return The job submitted
    */
-  def run(resetJobConfig: Boolean = true) = {
+  def run(resetJobConfig: Boolean = true): StreamJob = {
+    persist(resetJobConfig)
+    submit()
+  }
+
+  /**
+   * This function persist config in coordinator stream.
```
Comment (lhaiesp) [inline samza-core/src/main/scala/org/apache/samza/job/JobRunner.scala]: Besides creating and writing into coordinator stream, this method also creates diagnostics streams and potentially other stuff in the future. Suggest we provide more details in the java doc here. Also suggest a more generic method name such as "prepare()"

File: samza-core/src/main/scala/org/apache/samza/util/CoordinatorStreamUtil.scala [hunk 2]
```diff
@@ -37,13 +39,12 @@ object CoordinatorStreamUtil extends Logging {
     * Given a job's full config object, build a subset config which includes
     * only the job name, job id, and system config for the coordinator stream.
     */
-  def buildCoordinatorStreamConfig(config: Config) = {
+  def buildCoordinatorStreamConfig(config: Config): MapConfig = {
     val jobConfig = new JobConfig(config)
-    val buildConfigFactory = jobConfig.getCoordinatorStreamFactory();
+    val buildConfigFactory = jobConfig.getCoordinatorStreamFactory
```
Comment (mynameborat) [inline samza-core/src/main/scala/org/apache/samza/util/CoordinatorStreamUtil.scala]: nit: I'd recommend leaving the parentheses as is for convention. all java method call inside scala has explicit parentheses even in the absence of parameters.

File: samza-core/src/test/scala/org/apache/samza/util/TestCoordinatorStreamUtil.scala [hunk 3]
```diff
@@ -84,4 +84,19 @@ class TestCoordinatorStreamUtil {
...
+    addConfig.put("job.name", "test-job-name")
+    addConfig.put("job.id", "i001")
+    addConfig.put("job.coordinator.system", "samzatest")
+    addConfig.put("systems.samzatest.test","test")
+    addConfig.put("test.only","nothing")
+    addConfig.put("systems.samzatest.samza.factory", "org.apache.samza.system.MockSystemFactory")
+    val config = new MapConfig(addConfig)
+    val configMap = CoordinatorStreamUtil.buildCoordinatorStreamConfig(config)
+
+    CoordinatorStreamUtil.writeConfigToCoordinatorStream(configMap)
```
Comment (mynameborat) [inline samza-core/src/test/scala/org/apache/samza/util/TestCoordinatorStreamUtil.scala]: Can we verify if the configs are piped to the mock system producer? 
Now that you are adding tests, maybe add one for resetJobConfig to make sure the configs not part of the old configs get deleted?

**Comment by mynameborat** [comment]
Can you also fix the checkstyle issue
```
> Task :samza-core_2.11:checkstyleMain
[ant:checkstyle] /home/travis/build/apache/samza/samza-core/src/main/java/org/apache/samza/util/DiagnosticsUtil.java:163:12: error: Name 'DIAGNOSTICS_STREAM_ID' must match pattern '^[a-z][a-zA-Z0-9]*$'.
> Task :samza-core_2.11:checkstyleMain FAILED
```

File: samza-core/src/main/scala/org/apache/samza/util/CoordinatorStreamUtil.scala (hunk 2 above)
Comment (kw2542) [reply samza-core/src/main/scala/org/apache/samza/util/CoordinatorStreamUtil.scala]: reverted the change to keep consistency.

File: samza-core/src/test/scala/org/apache/samza/util/TestCoordinatorStreamUtil.scala (hunk 3 above)
Comment (kw2542) [reply samza-core/src/test/scala/org/apache/samza/util/TestCoordinatorStreamUtil.scala]: As discussed, in order for us to mock and verify, we need to mock multiple nested data, it is not worth the effort to do it for now, as the code are simply copied from existing code.

**Pull request merged**